src/main/java/appeng/util/item/AEItemStackRegistry.java

```diff
@@ -61,11 +78,31 @@ static synchronized AESharedItemStack getRegisteredStack( final @Nonnull ItemSta
 
 		if( ret == null )
 		{
-			ret = new AESharedItemStack( itemStack.copy() );
-			REGISTRY.put( ret, new WeakReference<>( ret ) );
+			final long newStackId = Platform.isClient() ? serverStackId : ++nextStackId;
+			ret = new AESharedItemStack( itemStack.copy(), newStackId );
+			registry().put( ret, new WeakReference<>( ret ) );
 		}
 		itemStack.setCount( oldStackSize );
 
 		return ret;
 	}
+
+	static synchronized AESharedItemStack getRegisteredStack( final long itemStackId )
+	{
+		if( itemStackId <= 0 )
+		{
+			return null;
+		}
+
+		// TODO: better search
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1135)